### PR TITLE
- fixed TN/NT bug in check fingerprints.

### PR DESCRIPTION
**Pull request by yfarjoun**
- added tests

I'm sorry for the need for this PR. The code was previous not properly tested and it was put in this project prematurely. I have now both added (good!) unit-tests, and separately tested this on a known LoH sample. Unfortunately I am not able to share the sample that I tested this on, but I see it is able to improve the LOD score when the tumor is identified correctly, and it does NOT do so when the normal is called 'tumor'. Also important to point out that if the sample is incorrect, none of the LOD scores are positive....

### CODE REVIEW

**Comment by yfarjoun** [comment]
@ktibbett who should review this? @eitanbanks ?

File: src/java/picard/fingerprint/FingerprintChecker.java [hunk 1]
```diff
@@ -138,10 +138,10 @@ public void setpLossofHet(final double pLossofHet) {
         SequenceUtil.assertSequenceDictionariesEqual(this.haplotypes.getHeader().getSequenceDictionary(),
                                                      iterator.getSequenceDictionary());
```
Comment (yfarjoun) [inline src/java/picard/fingerprint/FingerprintChecker.java]: mostly this is java8 stuff.

File: src/java/picard/fingerprint/FingerprintChecker.java [hunk 2]
```diff
@@ -586,23 +582,32 @@ public static MatchResults calculateMatchResults(final Fingerprint observedFp, f
             final HaplotypeProbabilities probs1 = observedFp.get(haplotypeBlock);
             if (probs1 == null) continue;
 
-            final HaplotypeProbabilityOfNormalGivenTumor normalizedProbs1 = new HaplotypeProbabilityOfNormalGivenTumor(probs1, pLoH);
-            final HaplotypeProbabilityOfNormalGivenTumor normalizedProbs2 = new HaplotypeProbabilityOfNormalGivenTumor(probs2, pLoH);
+            final HaplotypeProbabilityOfNormalGivenTumor prob1AssumingDataFromTumor = new HaplotypeProbabilityOfNormalGivenTumor(probs1, pLoH);
```
Comment (yfarjoun) [inline src/java/picard/fingerprint/FingerprintChecker.java]: better names

File: src/java/picard/fingerprint/FingerprintChecker.java [hunk 3]
```diff
@@ -586,23 +582,32 @@ public static MatchResults calculateMatchResults(final Fingerprint observedFp, f
...
             final Snp snp = probs2.getRepresentativeSnp();
             final DiploidGenotype externalGenotype = probs2.getMostLikelyGenotype(snp);
             final LocusResult lr = new LocusResult(snp,
-                                                   externalGenotype,
-                                                   probs1.getMostLikelyGenotype(snp),
-                                                   probs1.getObsAllele1(),
-                                                   probs1.getObsAllele2(),
-                                                   probs1.getLodMostProbableGenotype(),
-                                                   probs1.shiftedLogEvidenceProbabilityGivenOtherEvidence(probs2),
+                    externalGenotype,
```
Comment (yfarjoun) [inline src/java/picard/fingerprint/FingerprintChecker.java]: alignment and comments.

File: src/java/picard/fingerprint/FingerprintChecker.java [hunk 4]
```diff
@@ -586,23 +582,32 @@ public static MatchResults calculateMatchResults(final Fingerprint observedFp, f
...
+                    probs1.getLodMostProbableGenotype(),
+                    // expected sample log-likelihood
+                    probs1.shiftedLogEvidenceProbabilityGivenOtherEvidence(probs2),
+                    // random sample log-likelihood
                     probs1.shiftedLogEvidenceProbability(),
-                    probs2.shiftedLogEvidenceProbabilityGivenOtherEvidence(normalizedProbs1)-probs2.shiftedLogEvidenceProbability(),
-                    probs1.shiftedLogEvidenceProbabilityGivenOtherEvidence(normalizedProbs2)-probs1.shiftedLogEvidenceProbability());
+
+            // probs1 is tumor probs2 is normal, correct sample lod
+            prob1AssumingDataFromTumor.shiftedLogEvidenceProbabilityGivenOtherEvidence(probs2) -
```
Comment (yfarjoun) [inline src/java/picard/fingerprint/FingerprintChecker.java]: this line (604) looks like a change, but it actually isn't. happy to explain with a whiteboard.

File: src/java/picard/fingerprint/HaplotypeProbabilitiesFromGenotype.java [hunk 5]
```diff
@@ -24,8 +24,6 @@
 
 package picard.fingerprint;
 
-import static picard.util.MathUtil.*;
```
Comment (yfarjoun) [inline src/java/picard/fingerprint/HaplotypeProbabilitiesFromGenotype.java]: whitespaces only in this file.

File: src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java [hunk 6]
```diff
@@ -47,26 +47,31 @@ public HaplotypeProbabilityOfNormalGivenTumor(final HaplotypeProbabilities hpOfT
...
     public double[] getLikelihoods() {
-        final double[] asTumorLikelihoods = new double[3];
-        final double[] asNormalLikelihoods = hpOfTumor.getLikelihoods();
+        final double[] normalHaplotypeLikelihoods = new double[3];
+        final double[] tumorHaplotypeLikelihoods = hpOfTumor.getLikelihoods();
         for (final Genotype g_n : Genotype.values()) {
+            normalHaplotypeLikelihoods[g_n.v] = 0D;
             for (final Genotype g_t : Genotype.values()) {
-                asTumorLikelihoods[g_t.v] += asNormalLikelihoods[g_n.v] * transitionMatrix[g_n.v][g_t.v];
+                normalHaplotypeLikelihoods[g_n.v] += tumorHaplotypeLikelihoods[g_t.v] * transitionMatrix[g_n.v][g_t.v];
```
Comment (yfarjoun) [inline src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java]: This is the bug.

File: src/tests/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumorTest.java [hunk 7]
```diff
@@ -31,26 +29,27 @@
...
         testData.add(new Object[]{0.0, new double[]{0.3, 0.7, 0}, new double[]{0.3, 0.7, 0}});
 
-        //make sure that pLoH will not affect HOM likelihoods:
-        testData.add(new Object[]{0.1, new double[]{1, 0, 0}, new double[]{1, 0, 0}});
-        testData.add(new Object[]{0.2, new double[]{0, 0, 1}, new double[]{0, 0, 1}});
-        testData.add(new Object[]{0.3, new double[]{.3, 0, .7}, new double[]{.3, 0, .7}});
-
+        final double pLoh = 0.1;
         //see that non zero pLoH changes the likelihood of a HET site as expected:
-        testData.add(new Object[]{0.1, new double[]{0, 1, 0}, new double[]{.1/2, 1-0.1, .1/2}});
```
Comment (yfarjoun) [inline src/tests/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumorTest.java]: old tests were wrong...that didn't help.

File: src/tests/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 8]
```diff
@@ -26,4 +36,45 @@ public void testRandomSublist() throws Exception {
...
+        List<Object[]> listOfDoubles = new ArrayList<>();
+
+        for(int i=1;i<20; i++) {
+            listOfDoubles.add(new Object[]{i/40D});
+        }
+        return listOfDoubles.iterator();
+    }
+
+    @Test(dataProvider = "pLoH")
+    public void testMatchResults(final double pLoH) {
```
Comment (yfarjoun) [inline src/tests/java/picard/fingerprint/FingerprintCheckerTest.java]: new test. checks that if we test a hom fp against a het fp, it only give good results if the het is "normal" and the hom is "tumor"

File: src/java/picard/fingerprint/FingerprintChecker.java [hunk 9]
```diff
@@ -264,9 +264,7 @@ public IntervalList getLociToGenotype(final Collection<Fingerprint> fingerprints
             }
         }
 
-        intervals.sort();
-        intervals.unique();
-        return intervals;
+        return intervals.sorted().uniqued();
```
Comment (nh13) [inline src/java/picard/fingerprint/FingerprintChecker.java]: doesn't `uniqued` imply `sorted`?

File: src/java/picard/fingerprint/FingerprintChecker.java (hunk 4 above)
Comment (nh13) [inline src/java/picard/fingerprint/FingerprintChecker.java]: I'll take your word for it.

File: src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java [hunk 10]
```diff
@@ -1,7 +1,7 @@
 /*
  * The MIT License
  *
- * Copyright (c) 2010 The Broad Institute
+ * Copyright (c) 2016 The Broad Institute
```
Comment (nh13) [inline src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java]: `2010-2016`

File: src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java [hunk 11]
```diff
@@ -47,26 +47,31 @@ public HaplotypeProbabilityOfNormalGivenTumor(final HaplotypeProbabilities hpOfT
                 //This is P(g_t|g_n)
                 //tumor genotype are the columns.
                 {1,               0,        0},  //normal is hom_ref => tumor must be the same
-                {pLoH / 2, 1 - pLoH, pLoH / 2},  //normal is het     => tumor might transit
+                {pLoH / 2, 1 - pLoH, pLoH / 2},  //normal is het     => tumor may have transited
```
Comment (nh13) [inline src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java]: transitioned?

Comment (nh13) [inline src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java]: they're not taking a 🚌

File: src/tests/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 12]
```diff
@@ -26,4 +36,45 @@ public void testRandomSublist() throws Exception {
 
         Assert.assertEquals(FingerprintChecker.randomSublist(list, 2).size(), 2);
     }
+
+    @DataProvider(name="pLoH")
+    public Iterator<Object[]> pLohData(){
+        List<Object[]> listOfDoubles = new ArrayList<>();
```
Comment (nh13) [inline src/tests/java/picard/fingerprint/FingerprintCheckerTest.java]: `final`

File: src/tests/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 13]
```diff
@@ -26,4 +36,45 @@ public void testRandomSublist() throws Exception {
...
+
+        for(int i=1;i<20; i++) {
```
Comment (nh13) [inline src/tests/java/picard/fingerprint/FingerprintCheckerTest.java]: "s  p  a  c  e  s"

File: src/tests/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 14]
```diff
@@ -26,4 +36,45 @@ public void testRandomSublist() throws Exception {
...
+        MatchResults mr = FingerprintChecker.calculateMatchResults(fpObserved, fpExpected, 0.01, pLoH);
+
+        //make sure that it's more likely to be the same sample, if the observed is "tumor" and the expected is "normal"
+        Assert.assertTrue(mr.getLodTN() > mr.getLOD() );
+
+        //make sure that the regular LOD is negative (we're comparing a HET to a HOM)
+        Assert.assertTrue(mr.getLOD() < 0);
+
+        // make sure that it's more likely to be tumor/normal rather than normal/tumor
+        // (a hom normal isn't expected to be measured as a het in the tumor)
```
Comment (nh13) [inline src/tests/java/picard/fingerprint/FingerprintCheckerTest.java]: to space or not to space (after the `//`), that is the question.

File: src/tests/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 15]
```diff
@@ -26,4 +36,45 @@ public void testRandomSublist() throws Exception {
...
+        final HaplotypeProbabilities hpHomRef = new HaplotypeProbabilitiesFromGenotype(snp, hb, 1.0, 0.00001, 0.000000001);
+
+        //Expected is a het
+        fpExpected.add(hpHet);
+
+        //Observed is a hom, so possible scenario is that observed is tumor, and expected is normal
+        fpObserved.add(hpHomRef);
+
+        //get match results using pLOD
+        MatchResults mr = FingerprintChecker.calculateMatchResults(fpObserved, fpExpected, 0.01, pLoH);
```
Comment (nh13) [inline src/tests/java/picard/fingerprint/FingerprintCheckerTest.java]: final?

File: src/tests/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 16]
```diff
@@ -26,4 +36,45 @@ public void testRandomSublist() throws Exception {
...
+        //make sure that it's more likely to be the same sample, if the observed is "tumor" and the expected is "normal"
+        Assert.assertTrue(mr.getLodTN() > mr.getLOD() );
+
+        //make sure that the regular LOD is negative (we're comparing a HET to a HOM)
+        Assert.assertTrue(mr.getLOD() < 0);
+
+        // make sure that it's more likely to be tumor/normal rather than normal/tumor
+        // (a hom normal isn't expected to be measured as a het in the tumor)
+        Assert.assertTrue(mr.getLodTN() > mr.getLodNT());
+
```
Comment (nh13) [inline src/tests/java/picard/fingerprint/FingerprintCheckerTest.java]: extra newline

File: src/java/picard/fingerprint/FingerprintChecker.java (hunk 9 above)
Comment (yfarjoun) [inline src/java/picard/fingerprint/FingerprintChecker.java]: good point.

File: src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java (hunk 10 above)
Comment (yfarjoun) [inline src/java/picard/fingerprint/HaplotypeProbabilityOfNormalGivenTumor.java]: hmmm. The original version was not written in 2010...that's just a typo. I think it's fine as 2016...

**Comment by yfarjoun** [comment]
@jbloom22 bump!!

**Comment by jbloom22** [comment]
I'm sorry for taking so long given that, after a couple hours of digging around (and letting my eyes get used to Java and the testing framework), the code and variable names looks solid. If the tests pass, then I'm confident its doing the right thing. Is there a more numerical test you could do with regard to the test that catches the bug, or do you feel that's covered by the HaplotypeProbabilitiesTest, etc?

**Comment by yfarjoun** [comment]
the bug is definitely covered by the test (as I needed to change the test
to keep it passing...) if you have a suggestion for a test I'm happy to put
it in.

**Pull request merged**